### ISSUE TITLE
Add `at-rule-prelude-no-invalid`

### ISSUE THREAD

**Issue by jeddy3** [stylelint/stylelint#8152]
### What is the problem you're trying to solve?

Catch unknown preludes for at-rules.

For example:

```css
@property foo {}
```

As `@property` [expects](https://csstree.github.io/docs/syntax/#Atrule:property) a `<custom-property-name>` for its prelude:

```css
@property --foo {}
```

### What solution would you like to see?

Suggested blueprint:

- Name: `at-rule-prelude-no-invalid`
- Description: Disallow invalid preludes for at-rules.
- Message: `Unexpected invalid prelude "${prelude}" for at-rule "${at-rule}"`
- Primary option: `true`
- Secondary options: `ignoreAtRules: []`
- Category: Avoid errors > Invalid
- Included in sharable configs?: Yes, recommended when pieces are in place
- Expanded description: This rule considers preludes for at-rules defined within the CSS specifications to be valid.

We typically use `no-unknown` for things not in the specs. However, in hindsight, we probably should have used `no-invalid` and reserved `no-unknown` for user-defined things like animations and custom properties. I suggest we continue using `no-unknown` to be consistent with the existing rules.

The `ignoreAtRules: []` secondary option to avoid overlaps with existing rules like [`media-query-no-invalid`](https://stylelint.io/user-guide/rules/media-query-no-invalid/):

```json
{
  "rules": {
    "media-query-no-invalid": true,
    "at-rule-prelude-no-invalid": [true, { "ignoreAtRules": ["media"] }]
  }
}
```

**Comment by Mouvedia** [comment]
[As I explained](https://github.com/stylelint/stylelint/issues/8133#issuecomment-2486224673) in the parent issue, it should be named `at-rule-prelude-no-invalid`.

In your example—`@property foo {}`—`foo` is invalid.

i.e.
`@property` defines the custom property hence even if it was `--foo`—e.g. `@property --foo {}`—searching for a prior registration to know if it's known or not wouldn't make sense

> I suggest we continue using no-unknown to be consistent with the existing rules.

I disagree, we have 5 existing no-invalid rules as a precedent.

**Comment by jeddy3** [comment]
I think `@property foo {}` was a confusing example, sorry.

This rule would, like the existing [`declaration-property-value-no-unknown`](https://stylelint.io/user-guide/rules/declaration-property-value-no-unknown) one, use csstree to check all [at-rule and prelude pairs](https://csstree.github.io/docs/syntax/#discovery&q=ZGljdC5bdHlwZT0iQXRydWxlIl0%3D).

For example, it would flag:

```
@layer "unknown-string" {}
```

As `@layer` expects an `<ident>` rather than a `<string>` for its prelude.

**Comment by Mouvedia** [comment]
You are right that since it's for all at rules, it might be worth adding.
I still think it should be named `at-rule-prelude-no-invalid`.

**Comment by romainmenke** [comment]
Thank you for opening this issue @jeddy3!

----

Should we (in the future) consider deprecating `media-query-no-invalid`?
We'd have to check if this new rule covers the same range of issues.
But always nice if we can simplify things.

----

[note]

I don't currently have test coverage for atrule preludes in https://github.com/csstools/postcss-plugins/tree/main/packages/css-syntax-patches-for-csstree

**Comment by jeddy3** [comment]
It seems `media-query-no-invalid`  covers more issues.

For example, it'll catch these two:

```css
@media (width == 100px) {}
@media not(min-width: 300px) {}
```

Whereas csstree validator doesn't. I think that's because [it matches](https://csstree.github.io/docs/syntax/#Type:media-query-list) the:

- former as an `<any-value>` insides a `<general-enclosed>`
- latter as `<function-token> <any-value>` inside a `<general-enclosed>`

Although, I haven't fully grokked what's going on here. Do you have an insights?

**Comment by jeddy3** [comment]
SGTM. I named it to be consistent with the existing [declaration-property-value-no-unknown](https://stylelint.io/user-guide/rules/declaration-property-value-no-unknown) rule. In hindsight, we should've called that one `declaration-property-value-no-invalid`. It is what it is, though. We may as well get it right going forward and keep `declaration-property-value-no-unknown` as the outlier.
